repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,360 @@
+          val tokenValidationResult: Try[AuthResult] =
+            Option(datastore.get(s"$TOKEN_KEY_PREFIX$authToken").asInstanceOf[AuthResult]) map { validationResult =>
+              Success(validationResult)
+            } getOrElse {
+              //flatMap to unbox the Try[Try[TokenValidationResult]] so all Failure's are just packaged along
+              requestHandler.getAdminToken flatMap { adminToken =>
+                requestHandler.validateToken(adminToken, authToken) recoverWith {
+                  //Recover if the exception is an AdminTokenUnauthorizedException
+                  //This way we can specify however we want to what we want to do to retry.
+                  //Also it only retries ONCE! No loops or anything. Fails gloriously
+                  case unauth: AdminTokenUnauthorizedException =>
+                    //Clear the cache, call this method again
+                    datastore.remove(ADMIN_TOKEN_KEY)
+                    requestHandler.getAdminToken match {
+                      case Success(newAdminToken) =>
+                        requestHandler.validateToken(newAdminToken, authToken)
+                      case Failure(x) => Failure(IdentityAdminTokenException("Unable to reacquire admin token", x))
+                    }
+                }
+              }
+            }
+
+          tokenValidationResult match {
+            case Success(InvalidToken) =>
+              Reject(SC_UNAUTHORIZED)
+            case Success(validToken: ValidToken) =>
+              val uriTenantOption = requestHandler.extractTenant(request.getRequestURI)
+              val authorizedTenant = requestHandler.tenantAuthorization(uriTenantOption, validToken) match {
+                case Some(Success(tenantHeaderValues)) =>
+                  Pass(Map(OpenStackServiceHeader.TENANT_ID.toString -> tenantHeaderValues.mkString(",")))
+                case Some(Failure(e)) =>
+                  Reject(SC_UNAUTHORIZED, failure = Some(e))
+                case None =>
+                  Pass(Map.empty[String, String])
+              }
+
+              val endpoints = authorizedTenant match {
+                case Pass(headers) =>
+                  requestHandler.handleEndpoints(authToken, validToken) match {
+                    case Some(Success(endpointsData)) =>
+                      //If I'm configured to put the endpoints into a x-catalog do it
+                      if (config.getIdentityService.isSetCatalogInHeader) {
+                        // todo: don't check this flag twice
+                        val endpointsHeader = PowerApiHeader.X_CATALOG.toString -> Base64.encodeBase64String(endpointsData.endpointsJson.getBytes)
+                        Pass(headers + endpointsHeader)
+                      } else {
+                        Pass(headers)
+                      }
+                    case Some(Failure(x)) =>
+                      //Reject them with 403
+                      Reject(SC_FORBIDDEN, failure = Some(x))
```

Comment:
As i said I haven't gotten all the way through, but i'm assuming  Rejects will be looked at later, and if we are in delegating mode we'll populate the delegating header and doFilter, otherwise return a response immediately.